### ISSUE TITLE
Returning a function in `r.js` crashes the server

### ISSUE THREAD

**Issue by neumino** [rethinkdb/rethinkdb#4190]
Query:
```
r.expr([1,2,3,4,5]).filter(r.js('(function(row) { return function() {} })'))
```

Backtrace:
```
error: Error in src/extproc/js_runner.cc at line 230:
error: Guarantee failed: [cache_res.second] js_runner_t cached two functions with the same source
error: Backtrace:
addr2line: 'rethinkdb': No such file
addr2line: 'rethinkdb': No such file
error: Tue May  5 22:37:13 2015
       
       1: backtrace_t::backtrace_t() at 0x97e280 (rethinkdb)
       2: format_backtrace(bool) at 0x97e5f9 (rethinkdb)
       3: report_fatal_error(char const*, int, char const*, ...) at 0xb8e68d (rethinkdb)
       4: js_runner_t::cache_id(unsigned long, std::string const&) at 0x8d4b1e (rethinkdb)
       5: js_runner_t::call(std::string const&, std::vector<ql::datum_t, std::allocator<ql::datum_t> > const&, js_runner_t::req_config_t const&) at 0x8d5590 (rethinkdb)
       6: ql::js_func_t::call(ql::env_t*, std::vector<ql::datum_t, std::allocator<ql::datum_t> > const&, ql::eval_flags_t) const at 0xa5290f (rethinkdb)
       7: ql::js_func_t::filter_helper(ql::env_t*, ql::datum_t) const at 0xa51da2 (rethinkdb)
       8: ql::func_t::filter_call(ql::env_t*, ql::datum_t, counted_t<ql::func_t const>) const at 0xa51252 (rethinkdb)
       9: ql::filter_trans_t::lst_transform(ql::env_t*, std::vector<ql::datum_t, std::allocator<ql::datum_t> >*, ql::datum_t const&) at 0x9fa721 (rethinkdb)
       10: ql::ungrouped_op_t::operator()(ql::env_t*, std::map<ql::datum_t, std::vector<ql::datum_t, std::allocator<ql::datum_t> >, optional_datum_less_t, std::allocator<std::pair<ql::datum_t const, std::vector<ql::datum_t, std::allocator<ql::datum_t> > > > >*, ql::datum_t const&) at 0x9f8ff3 (rethinkdb)
       11: ql::eager_datum_stream_t::next_grouped_batch(ql::env_t*, ql::batchspec_t const&, std::map<ql::datum_t, std::vector<ql::datum_t, std::allocator<ql::datum_t> >, optional_datum_less_t, std::allocator<std::pair<ql::datum_t const, std::vector<ql::datum_t, std::allocator<ql::datum_t> > > > >*) at 0xad4a03 (rethinkdb)
       12: ql::eager_datum_stream_t::next_batch_impl(ql::env_t*, ql::batchspec_t const&) at 0xad4e5d (rethinkdb)
       13: ql::datum_stream_t::next_batch(ql::env_t*, ql::batchspec_t const&) at 0xad27b2 (rethinkdb)
       14: ql::datum_stream_t::next(ql::env_t*, ql::batchspec_t const&) at 0xad2fdb (rethinkdb)
       15: ql::array_datum_stream_t::next(ql::env_t*, ql::batchspec_t const&) at 0xad3205 (rethinkdb)
       16: ql::eager_datum_stream_t::as_array(ql::env_t*) at 0xad2c3b (rethinkdb)
       17: ql::array_datum_stream_t::as_array(ql::env_t*) at 0xade748 (rethinkdb)
       18: ql::val_t::val_t(ql::env_t*, counted_t<ql::datum_stream_t>, ql::protob_t<Backtrace const>) at 0xaeffd7 (rethinkdb)
       19: scoped_ptr_t<ql::val_t> make_scoped<ql::val_t, ql::env_t*, counted_t<ql::datum_stream_t>, ql::protob_t<Backtrace const> const&>(ql::env_t*&&, counted_t<ql::datum_stream_t>&&, ql::protob_t<Backtrace const> const&) at 0xa75318 (rethinkdb)
       20: ql::filter_term_t::eval_impl(ql::scope_env_t*, ql::args_t*, ql::eval_flags_t) const at 0xa9dd09 (rethinkdb)
       21: ql::op_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at 0xad0734 (rethinkdb)
       22: ql::runtime_term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at 0xacd33d (rethinkdb)
       23: ql::query_cache_t::ref_t::run(ql::env_t*, Response*) at 0xae80f1 (rethinkdb)
       24: ql::query_cache_t::ref_t::fill_response(Response*) at 0xae9a7e (rethinkdb)
       25: ql::run(ql::query_id_t&&, ql::protob_t<Query>, Response*, ql::query_cache_t*, signal_t*) at 0xacd6e0 (rethinkdb)
       26: rdb_query_server_t::run_query(ql::query_id_t&&, ql::protob_t<Query> const&, Response*, ql::query_cache_t*, signal_t*) at 0xaf2c2f (rethinkdb)
       27: void save_exception<void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::{lambda(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*)#1}::operator()(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*) const::{lambda()#1}>(std::__exception_ptr::exception_ptr*, std::string*, cond_t*, void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::{lambda(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*)#1}::operator()(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*) const::{lambda()#1}&&) at 0x9284b1 (rethinkdb)
       28: void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::{lambda(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*)#1}::operator()(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, signal_t*) const at 0x928ab2 (rethinkdb)
       29: coro_pool_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > >::worker_run(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, auto_drainer_t::lock_t) at 0x9266a4 (rethinkdb)
       30: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (coro_pool_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > >::*)(std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, auto_drainer_t::lock_t)> (coro_pool_t<std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > > >*, std::_List_iterator<std::pair<ql::query_id_t, ql::protob_t<Query> > >, auto_drainer_t::lock_t)> >::run_action() at 0x9261b4 (rethinkdb)
       31: coro_t::run() at 0xbb13e8 (rethinkdb)
error: Exiting.
Trace/breakpoint trap (core dumped)
```

**Comment by mlucy** [comment]
Urkh.  Thanks for catching that.

Assigning to @Tryneus because `vc-annotate` makes it look like he wrote this code, but feel free to reassign if it's misleading me.

**Comment by Tryneus** [comment]
Fix for this is up in review 2905.

**Comment by danielmewes** [comment]
@Tryneus the review seems to be donie. Is this in v2.0.x?

**Comment by mlucy** [comment]
@danielmewes -- I approved the review late last week when @Tryneus had already left for vacation, so I would guess not.

**Comment by Tryneus** [comment]
The fix has been approved and merged into `next` in commit 8f8a240, and cherry-picked into `v2.0.x` in commit e117478.  Will be in release 2.0.3.
